src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -152,188 +164,159 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
-                                    // Less than min possible headerSpan of 5 bytes a:b\r\n
-                                    // Not parsable as a valid name:value header pair.
-                                    RejectRequestHeader(span[..length]);
-                                }
-
-                                // Read the header successfully, skip the reader forward past the headerSpan.
-                                span = span.Slice(length);
-                                reader.Advance(length);
-                            }
-                            else
-                            {
-                                // No enough data, set length to 0.
-                                length = 0;
-                            }
+                        if (length != 0 && !TryTakeSingleHeader(handler, span[..length]))
+                        {
+                            // Sequence needs to be CRLF and not contain an inner CR not part of terminator.
+                            // Not parsable as a valid name:value header pair.
+                            RejectRequestHeader(span[..lineLength]);
                         }
+
+                        // Read the header successfully, skip the reader forward past the headerSpan.
+                        span = span[lineLength..];
+                        reader.Advance(lineLength);
                     }
 
                     // End found in current span
                     if (length > 0)
                     {
+                        length = -1;
                         continue;
                     }
 
-                    // We moved the reader to look ahead 2 bytes so rewind the reader
-                    if (readAhead > 0)
-                    {
-                        reader.Rewind(readAhead);
-                    }
+                    // Load next header line to parse as a span
+                    span = ParseMultiSpanHeader(ref reader, out length);
 
-                    length = ParseMultiSpanHeader(handler, ref reader);
-                    if (length < 0)
+                    // If there any remaining line?
+                    if (length == -1 && span.Length == 0)
                     {
-                        // Not there
                         return false;
                     }
-
-                    reader.Advance(length);
-                    // As we crossed spans set the current span to default
-                    // so we move to the next span on the next iteration
-                    span = default;
                 }
             }
 
             return false;
         }
 
-        private int ParseMultiSpanHeader(TRequestHandler handler, ref SequenceReader<byte> reader)
+        // Returns the length of the line terminator (CRLF = 2, LF = 1)
+        // If no valid EOL is detected then -1
+        private int ParseHeaderLineEnd(ReadOnlySpan<byte> headerSpan, int length)
```

Comment:
Consider adding an assert and/or comment that `length` is nonnegative coming into this method.